### ISSUE TITLE
Very slow decoding of paletted PNG images

### ISSUE THREAD

**Issue by Shnatsel** [image-rs/image-png#393]
[`zune-png` benchmarks](https://etemesi254.github.io/assets/criterion/png_%20PNG%20decoding%20palette%20image/report/index.html) show that the `png` crate is much slower than other decoders on indexed images - a whopping 3x slower than `zune-png`.

[CPU profile](https://share.firefox.dev/3MqTcun) shows that 71% of the time is spent in `png::utils::unpack_bits`, specifically [this hot loop](https://github.com/image-rs/image-png/blob/2f53fc40b91a5e1d0ad1801b746c04a7fe1d8603/src/utils.rs#L35-L37).

The code is full of indexing and doesn't look amenable to autovectorization. I think the entire function will have to be rewritten; it's probably a good idea to copy `zune-png` here.

**Comment by Shnatsel** [comment]
Measured on `v0.17.8-rc` with this image: [speed_bench_palette](https://user-images.githubusercontent.com/291257/229297591-30e20b14-62f9-42b9-978a-563b66db5578.png)

**Comment by Shnatsel** [comment]
`zune-png` uses vectorization-friendly implementation of expansion, which explains the difference in performance. It can be found here: https://github.com/etemesi254/zune-image/blob/60fbeef1e58a7dec08e27bf2ed190e061f27cb45/zune-png/src/utils.rs#L1

**Comment by fintelia** [comment]
The version in this crate handles paletted images with 1/2/4-bit indices, whereas `zune-png` seems not to. However, probably the overwhelming majority of images use 8-bit indices... so it is probably worth special casing it.

**Comment by Shnatsel** [comment]
Based on [this comment](https://github.com/etemesi254/zune-image/blob/60fbeef1e58a7dec08e27bf2ed190e061f27cb45/zune-png/src/utils.rs#L169-L171) I understand `zune-png` expands the bits to bytes first, and then resolves those to the correct palette values in another pass.

So it does handle lower bit depth indices correctly, but they're expanded to 8 bits before use.

**Comment by fintelia** [comment]
This reminded me that I started working on more refactoring which included changing the transformations to not operate in place. Just created https://github.com/image-rs/image-png/pull/396 with the current state

**Comment by anforowicz** [comment]
Looking at top-500 website, paletted images account for 35% of PNG images.  Source:
* https://github.com/image-rs/image-png/discussions/416#discussioncomment-7436871
* https://docs.google.com/spreadsheets/d/1CEgr4ttd2agOZi-MFOjqJGqfDVFUKDp-zKbrF-8Xwyk/edit?usp=sharing (see the blue part of the "image properties" sheet)

OTOH so far I only had negative results when trying to improve paletted images: https://github.com/image-rs/image-png/discussions/416#discussioncomment-7590601

**Comment by Shnatsel** [comment]
It's the bit unpacking hot loop that's slowing things down. `zune-png` has already figured out a much more performant way to do it that gets auto-vectorized: https://github.com/etemesi254/zune-image/blob/54cc956ccc01ea942456c0dcebf8d97bda614666/crates/zune-png/src/utils.rs#L217-L317

It should be fairly straightforward to integrate into the `png` crate.

**Comment by okaneco** [comment]
> It's [the bit unpacking hot loop](https://github.com/image-rs/image-png/blob/2f53fc40b91a5e1d0ad1801b746c04a7fe1d8603/src/utils.rs#L35-L37) that's slowing things down.
 
That code was changed in #405 and [looks like this now](https://github.com/image-rs/image-png/blob/b00fb53e18d8cde6485d20623e0e058e3f88e917/src/utils.rs#L6-L51).

The distinction I'd make is that it's not the bit-unpacking as much as retrieving colors from the lookup table and filling those into the buffer. Trying to do it in 2 passes ended up slower than the current behavior of unpacking and then doing the lookup in 1 pass. The code didn't seem to get autovectorized even using `chunks_exact` at the time.

I'm not sure that it's straightforward or easy to get similar results without larger architectural changes as mentioned in https://github.com/image-rs/image-png/discussions/416#discussioncomment-7590601 where different several methods were tried unsuccessfully.

**Comment by Shnatsel** [comment]
It is clearly possible to do better here still, as evidenced by `zune-png`. Even with the latest code it is 50% faster on [a big paletted image](https://user-images.githubusercontent.com/291257/229297591-30e20b14-62f9-42b9-978a-563b66db5578.png). Perhaps its code is worth a closer look.

**Comment by anforowicz** [comment]
> It is clearly possible to do better here still, as evidenced by `zune-png`. Even with the latest code it is 50% faster on [a big paletted image](https://user-images.githubusercontent.com/291257/229297591-30e20b14-62f9-42b9-978a-563b66db5578.png). Perhaps its code is worth a closer look.

FWIW, I've tried running `perf record` / `perf report` on the linked image and it seems that most time is spent _below_ the `png` crate (i.e. not in `fn expand_paletted` in `png/src/decoder/mod.rs` which is implemented in the `png` crate).  I understand that inlining and other optimizations may make the `perf report` somewhat inaccurate/crude, but AFAIU things like `fdeflate::decompress::Decompressor::read` should only accidentally include things inlined from dependencies of `fdeflate` (like the `simd-adler32` crate), but not from clients of `fdeflate` (like the `png` crate).  I guess LTO can potentially deduplicate code across `fdeflate` and `png` and continue attributing profiling samples to `fdeflate`, but this seems unlikely for `fn expand_paletted`.

Maybe this means that the observed runtime delta (compared to `zune-png`) comes from other areas?

For the record, here are the `perf report` results I got for that paletted image:

```
  55.34%  decoder-1b5b1e4  decoder-1b5b1e46cb85776a  [.] fdeflate::decompress::Decompressor::read                                                                                                                                            
  17.43%  decoder-1b5b1e4  decoder-1b5b1e46cb85776a  [.] <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter                                                                                                    
  12.61%  decoder-1b5b1e4  libc-2.31.so              [.] 0x00000000001602f2                                                                                                                                                                  
   3.75%  decoder-1b5b1e4  [kernel.kallsyms]         [k] clear_page_erms                                                                                                                                                                     
   2.56%  decoder-1b5b1e4  [kernel.kallsyms]         [k] read_hpet                                                                                                                                                                           
   2.33%  decoder-1b5b1e4  decoder-1b5b1e46cb85776a  [.] crc32fast::specialized::pclmulqdq::calculate                                                                                                                                        
   1.29%  decoder-1b5b1e4  decoder-1b5b1e46cb85776a  [.] fdeflate::decompress::Decompressor::build_tables                                                                                                                                    
   0.70%  decoder-1b5b1e4  decoder-1b5b1e46cb85776a  [.] <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter                                                                                                    
   0.49%  decoder-1b5b1e4  decoder-1b5b1e46cb85776a  [.] simd_adler32::imp::avx2::imp::update_imp                                                       
...
```

**Comment by Shnatsel** [comment]
`debug=true` in `Cargo.toml` combined with `perf --call-graph=dwarf` should attribute inlined code correctly. I'll profile the decoding of this image later and post the results.

**Comment by Shnatsel** [comment]
You need to call `decoder.set_transformations(png::Transformations::EXPAND)` to activate conversion from the palette indices to RGBA. If this is not set, the conversion from palette indices to RGB is never performed, and will not show up on the profile at all.

If you set that decoder option, then 40% of the total decoding time is spent in `expand_paletted` - here's the profile: https://share.firefox.dev/422gfkL

It is measured on this image, meant to be a a more realistic use case for indexed PNG: [Stadt_Onex_2021_indexed_oxipng](https://github.com/image-rs/image-png/assets/291257/ced998d0-a608-4b97-af36-649393769a68), originally sourced [from wikimedia](https://commons.wikimedia.org/wiki/File:Stadt_Onex_2021.png) and converted to paletted image by myself.

It is possible that Chromium prefers to resolve the indices themselves using the palette chunk instead of getting RGB or RGBA output from the decoder, in which case this is bottleneck is not relevant to Chromium.

This hot loop is where 40% of the decoding time for this image is spent: 

https://github.com/image-rs/image-png/blob/b00fb53e18d8cde6485d20623e0e058e3f88e917/src/decoder/mod.rs#L872-L879

Code used for measurement: https://gist.github.com/Shnatsel/ec75b05205ef55c96b9d256fdc36c2ec

**Comment by anforowicz** [comment]
> You need to call `decoder.set_transformations(png::Transformations::EXPAND)` to activate conversion from the palette indices to RGBA. If this is not set, the conversion from palette indices to RGB is never performed, and will not show up on the profile at all.

Ooops.  You're right, my bad.  I knew about that but forgot :-(.

> If you set that decoder option, then 40% of the total decoding time is spent in `expand_paletted` - here's the profile: https://share.firefox.dev/422gfkL

Thanks for sharing.

> It is measured on this image, meant to be a a more realistic use case for indexed PNG: [Stadt_Onex_2021_indexed_oxipng](https://github.com/image-rs/image-png/assets/291257/ced998d0-a608-4b97-af36-649393769a68), originally sourced [from wikimedia](https://commons.wikimedia.org/wiki/File:Stadt_Onex_2021.png) and converted to paletted image by myself.

Ack.

> It is possible that Chromium prefers to resolve the indices themselves using the palette chunk instead of getting RGB or RGBA output from the decoder, in which case this is bottleneck is not relevant to Chromium.

No, Chromium also wants to expand into RGBA (at a high-level, there are some other unimportant details like BGRA and/or alpha-premul).

> This hot loop is where 40% of the decoding time for this image is spent:
> 
> https://github.com/image-rs/image-png/blob/b00fb53e18d8cde6485d20623e0e058e3f88e917/src/decoder/mod.rs#L872-L879
> 
> Code used for measurement: https://gist.github.com/Shnatsel/ec75b05205ef55c96b9d256fdc36c2ec

I've tried to replicate your results with the code at https://github.com/anforowicz/image-png/tree/palette-measurements, but I got different results.  I am very curious what may be causing the difference.

Repro steps:

1. Run the following

    ```
    # booted with isolcpus=nohz,domain,4-5
    # (I forgot to disable TurboBoost and move kernel work off of cores 4-5,
    # but hopefully this shouldn't matter for comparing how much a function takes in a profile)
    $ git reset --hard
    $ git checkout 1f57fd0cd050d1fa8ed43c3d94c781e6057aa483 # this is the `palette-measurements` branch
    $ rustup run stable rustc --version
    rustc 1.75.0 (82e1608df 2023-12-21)
    $ rustup run stable cargo build --bench=decoder --release   
    ...
    $ taskset --cpu-list 4-5 nice -n -19 sudo perf record --call-graph=dwarf target/release/deps/decoder-01bdb75be1fb05ce --bench --profile-time 10 big-palletted
    $ sudo perf script -F +pid > /tmp/test.perf
    ...
    ```

2. Upload the profile to https://profiler.firefox.com/ (this is great - thanks for teaching me about this tool)
3. Share the profile

In the https://share.firefox.dev/3tOf0ZM that got, I can see the `png::decoder::expand_paletted::_$u7b$$u7b$closure$u7d$$u7d$::hfd10b92927fc4892` closure in the call tree when using inverted call stack, but it only takes 0.1% of the whole profile

**Comment by Shnatsel** [comment]
If this is a benchmark, you need to set `debug = true` for the `bench` profile and not just the `release` one. Cargo uses the `bench` profile when compiling benchmarks:

> Cargo has 4 built-in profiles: dev, release, test, and bench. 

([from Cargo docs](https://doc.rust-lang.org/cargo/reference/profiles.html))

---

Speaking of tools, [samply](https://crates.io/crates/samply) is amazing. Try this:

```
cargo install samply
echo '0' | sudo tee /proc/sys/kernel/perf_event_paranoid  # to be able to profile as non-root
samply record /path/to/binary arg1 arg2
```

Its processed results are more accurate than `perf`, it produces them much faster than `perf script`, it opens Firefox Profiler automatically, and you can double-click flame graph bars to open the code view while Samply is running to see the samples attributed to both lines of code and assembly instructions.

**Comment by Shnatsel** [comment]
Ah, I've found your problem: there are two different `Decoder` instances in the benchmarks and you applied the transform to one but not the other. The patched one:
https://github.com/anforowicz/image-png/blob/1f57fd0cd050d1fa8ed43c3d94c781e6057aa483/benches/decoder.rs#L60-L61
is not performing the actual decoding, this one is:
https://github.com/anforowicz/image-png/blob/1f57fd0cd050d1fa8ed43c3d94c781e6057aa483/benches/decoder.rs#L69
And you forgot to set the EXPAND transformation on it as well.

**Comment by anforowicz** [comment]
> Ah, I've found your problem: there are two different `Decoder` instances in the benchmarks and you applied the transform to one but not the other.

Ugh.  You're right.  Thank you for pointing this out. That's rather embarassing (since I've abandoned my earlier attempts at improving `expand_paletted` because of this incorrect approach to benchmarking).

For now I've opened https://github.com/image-rs/image-png/pull/453 to cover the `expand_paletted` in the end-to-end benchmarks (adding one of the images you've pointed out above).

I'll try to find some time next week to also work on:

* Function-level benchmarks of `expand_paletted` (https://github.com/anforowicz/image-png/tree/palette-benchmarks-func).  The first step will be figuring out why the end-to-end benchmarks move, even though this was supposed to be a performance-neutral refactoring that just hides `expand_paletted` in a module that exposes a simple, testable, benchable API.
* Actual improvements.  I see around -13% improvement for the new end-to-end benchmark image with https://github.com/anforowicz/image-png/tree/palette-combined-rgba, but currently Transparency sees a 2-3% regression.
    - I've tried looking at https://github.com/etemesi254/zune-image/blob/cd696f8f23dddee14063e66c6a728163ad0e6e2b/crates/zune-bmp/src/decoder.rs#L807.  It seems quite similar to what I came up with in the commit above.
    - There are probably opportunities to read/write 4 bytes at a time (rather than just 3).  This may require side-stepping `utils::unpack_bits`.
    - And there may be opportunities for using `std::simd` maybe?

**Comment by Shnatsel** [comment]
> The first step will be figuring out why the end-to-end benchmarks move, even though this was supposed to be a performance-neutral refactoring that just hides `expand_paletted` in a module that exposes a simple, testable, benchable API.

Probably because of inlining. [More on inlining in Rust](https://matklad.github.io/2021/07/09/inline-in-rust.html)

You can use https://github.com/pacak/cargo-show-asm to inspect the generated assembly, or just `objdump -d` to keep it old school. You're mostly interested in whether a given function is present as a separate function or not (inlined ones disappear from the listing).

**Comment by Shnatsel** [comment]
Fixed by #462
